packages/smooth_app/lib/background/background_task_language_refresh.dart

```diff
@@ -0,0 +1,136 @@
+    result[_jsonTagExcludeBarcodes] = excludeBarcodes;
+    return result;
+  }
+
+  static const OperationType _operationType = OperationType.languageRefresh;
+
+  static Future<void> addTask(
+    final LocalDatabase localDatabase, {
+    final List<String> excludeBarcodes = const <String>[],
+  }) async {
+    final String uniqueId = await _operationType.getNewKey(
+      localDatabase,
+    );
+    final BackgroundTask task = _getNewTask(
+      uniqueId,
+      excludeBarcodes,
+    );
+    await task.addToManager(localDatabase);
+  }
+
+  @override
+  (String, AlignmentGeometry)? getFloatingMessage(
+          final AppLocalizations appLocalizations) =>
+      null;
+
+  static BackgroundTask _getNewTask(
+    final String uniqueId,
+    final List<String> excludeBarcodes,
+  ) =>
+      BackgroundTaskLanguageRefresh._(
+        processName: _operationType.processName,
+        uniqueId: uniqueId,
+        stamp: ';languageRefresh',
+        excludeBarcodes: excludeBarcodes,
+      );
+
+  @override
+  Future<void> preExecute(final LocalDatabase localDatabase) async {}
+
+  @override
+  bool get hasImmediateNextTask => true;
+
+  /// Number of products to download each time.
+  static const int _pageSize = 20;
+
+  @override
+  Future<void> execute(final LocalDatabase localDatabase) async {
+    final DaoProduct daoProduct = DaoProduct(localDatabase);
+    final OpenFoodFactsLanguage language = ProductQuery.getLanguage();
+    final List<String> barcodes = await daoProduct.getTopProductsToTranslate(
+      language,
+      limit: _pageSize,
+      excludeBarcodes: excludeBarcodes,
+    );
+    if (barcodes.isEmpty) {
+      return;
+    }
+    final SearchResult searchResult = await OpenFoodAPIClient.searchProducts(
+      getUser(),
+      ProductSearchQueryConfiguration(
+        fields: ProductQuery.fields,
+        parametersList: <Parameter>[
+          const PageSize(size: _pageSize),
+          const PageNumber(page: 1),
+          BarcodeParameter.list(barcodes),
+        ],
+        language: language,
+        country: ProductQuery.getCountry(),
+        version: ProductQuery.productQueryVersion,
+      ),
+      uriHelper: uriProductHelper,
+    );
+    if (searchResult.products == null || searchResult.count == null) {
+      throw Exception('Cannot refresh language');
+    }
+
+    // save into database and refresh all visible products.
+    await daoProduct.putAll(searchResult.products!, language);
+    localDatabase.upToDate.setLatestDownloadedProducts(searchResult.products!);
+
+    // Next page
+    final List<String> newExcludeBarcodes = <String>[];
+    // we keep the old "excluded" barcodes,...
+    newExcludeBarcodes.addAll(excludeBarcodes);
+    // ...add the new barcodes...
+    newExcludeBarcodes.addAll(barcodes);
+    // ...and remove barcodes we actually found on the server.
+    for (final Product product in searchResult.products!) {
+      newExcludeBarcodes.remove(product.barcode);
```

Comment:
Say we try to refresh a product that was removed in the time before the refresh, then we'll have a lopped until the Cannot refresh language exception

Let's remove the ones we queried, not the ones we received

Reply:
> Let's remove the ones we queried, not the ones we received

First we add all the barcodes we expected, and then remove all the barcodes we received. In the end we exclude the barcodes we expected but haven't received.
Here we're dealing with an exclusion barcode list: we need to tell the next task "I have already tried this barcode but received nothing".
```dart
    // ...add the new barcodes...
    newExcludeBarcodes.addAll(barcodes);
    // ...and remove barcodes we actually found on the server.
    for (final Product product in searchResult.products!) {
      newExcludeBarcodes.remove(product.barcode);
```

Reply:
@M123-dev Just to clarify a bit:
1. we run a query in the local database - "which products are not yet in the expected language?"
2. we maintain a list of "excluded" barcodes - barcodes we already tried to download but were not returned by the server
3. the products returned by the server won't get listed again by the (1) query, as they are now flagged with the expected language
4. and then we go back to (1) query in the next background task, looking for products not in the expected language and not queried yet.